### ISSUE TITLE
Date in Admin displays with mixed up character between time and AM/PM instead of a space when using PHP 8.3.

### ISSUE THREAD

**Issue by piloujp** [zencart/zencart#6229]
A mixed up character displays between time and AM/PM instead of a space when using PHP 8.3 on Admin page date.

Zen Cart version 1.5.8a, V2.0.0-alpha1 release and last GitHub files version too, all freshly installed.
PHP version: 8.3.2
Problem does not appear on PHP 8.1 and 8.2, only on PHP 8.3
System is Windows 11
Tested with different browsers with no influence.

![Capture_ecran2024-02-15](https://github.com/zencart/zencart/assets/117885587/6334c9ac-361a-45c9-b65b-4f9caed7a264)

Problem is coming from the hard coded conversion from ISO-8859-1 to UTF-8 in '..\admin\includes\header.php' at line 215, even if date provided is already UTF-8. Recent Windows systems (Windows 10, 11, Server 2022 and at least for PHP binaries Server 2019) support UTF-8 and don't need this conversion... Problem already mentioned in an old discussion about zcDate class.
`        echo((strtoupper(substr(PHP_OS, 0, 3)) == 'WIN') ? iconv('ISO-8859-1', 'UTF-8', $zcDate->output(ADMIN_NAV_DATE_TIME_FORMAT, time())) : $zcDate->output(ADMIN_NAV_DATE_TIME_FORMAT, time())); //windows does not "do" UTF-8...so a manual conversion is necessary`
An easy fix would be to remove this conversion that should only be applied on Windows Server 2016 and older using English language. The conversion will probably make problems on any system using encoding other than ISO-8859-1. With non-alphabetic characters sets, date will be surely mixed up.
It is possible to improve this conversion like this too:
`        echo((strtoupper(substr(PHP_OS, 0, 3)) == 'WIN') ? mb_convert_encoding($zcDate->output(ADMIN_NAV_DATE_TIME_FORMAT, time()), 'UTF-8') : $zcDate->output(ADMIN_NAV_DATE_TIME_FORMAT, time())); //windows does not "do" UTF-8...so a manual conversion is necessary`
Like this, it is not hard coded anymore.

**Comment by drbyte** [comment]
Most likely will also need to check that the `mb_convert_encoding()` function is available, because the `mbstring` extension is not always enabled (it's disabled by default in PHP: https://www.php.net/manual/en/mbstring.installation.php ).

**Comment by drbyte** [comment]
Suggestion: Why not just use a standard date format? 

```date(DATE_RFC7231)```
<img width="196" alt="Screen Shot 2024-02-15 at 5 22 14 PM" src="https://github.com/zencart/zencart/assets/404472/79690d75-257a-413a-bfaa-976c2d040aee">

Much less code, and no complex translations based on UTF encodings having to re-encode between locales, etc.

```diff
        <?php
-        echo((strtoupper(substr(PHP_OS, 0, 3)) == 'WIN') ? iconv('ISO-8859-1', 'UTF-8', $zcDate->output(ADMIN_NAV_DATE_TIME_FORMAT, time())) : $zcDate->output(ADMIN_NAV_DATE_TIME_FORMAT, time())); //windows does not "do" UTF-8...so a manual conversion is necessary
-        echo '&nbsp;' . date("O", time()) . ' GMT';  // time zone
+        echo date(DATE_RFC7231);
        echo '&nbsp;[' . $_SERVER['REMOTE_ADDR'] . ']'; // current admin user's IP address
```

**Comment by piloujp** [comment]
Excellent idea! Except format you choose displays local time with 'GMT' hard coded at the end.  It will be correct only if you are in Greenwich meridian. Hopefully, there are other formats:

`echo date(DATE_RFC2822);`
![CaptureRFC2822](https://github.com/zencart/zencart/assets/117885587/68754429-fd9b-4256-95ab-1074ebeced4e)

Or better, a custom sting:

`echo date("D, d M Y H:i:s \\G\\M\\TO");`
![CaptureDateString](https://github.com/zencart/zencart/assets/117885587/65188f75-8a9b-4029-8125-9804b87fdfec)

Like this, it looks exactly as before.

**Comment by piloujp** [comment]
Actually, original format was more like this, except 12/24H time display:

`echo date("l d M Y H:i");`
![CaptureDate](https://github.com/zencart/zencart/assets/117885587/e3336549-b3cb-44aa-afe5-0ff637e0789d)

**Comment by drbyte** [comment]
"GMT" was already hard coded in ZC.

**Comment by piloujp** [comment]
#"GMT" was already hard coded in ZC.
Yes, but with time zone shift displayed before. Look what I get with old format and RFC7231:
Actual ZC date:
![Capture_Date_old_ZC](https://github.com/zencart/zencart/assets/117885587/1771da7f-fe8b-4b54-8a06-439b678ecddc)

RFC7231:
![Capture_RFC7231](https://github.com/zencart/zencart/assets/117885587/b3a69653-cbfb-4ed5-a645-fe6130f86778)

For any reason, RFC7231 drop time zone information make you think displayed time is UTC.

**Comment by piloujp** [comment]
About `ADMIN_NAV_DATE_TIME_FORMAT`, at first I liked the idea of dropping it but after few tests including Japanese version, I think keeping it is better. It will make it easier to use other languages and their date format.
Code change could be like this:
From
```
        echo((strtoupper(substr(PHP_OS, 0, 3)) == 'WIN') ? mb_convert_encoding($zcDate->output(ADMIN_NAV_DATE_TIME_FORMAT, time()), 'UTF-8') : $zcDate->output(ADMIN_NAV_DATE_TIME_FORMAT, time())); //windows does not "do" UTF-8...so a manual conversion is necessary
        echo '&nbsp;' . date("O", time()) . ' GMT';  // time zone
```
To
`        echo date(ADMIN_NAV_DATE_TIME_FORMAT);`
in 'header.php' file, then change ADMIN_NAV_DATE_TIME_FORMAT to this for English:
`    'ADMIN_NAV_DATE_TIME_FORMAT' => 'l d M Y H:i \\G\\M\\T P',`
Which displays this:
![Capture_english_date](https://github.com/zencart/zencart/assets/117885587/47457b55-caf5-4a3f-a0ab-0379682355a5)

Or this for Japanese as an example:
`    'ADMIN_NAV_DATE_TIME_FORMAT' => 'Y年m月d日 H:i \\G\\M\\T P',`
Which displays this:
![Capture_japanese_date](https://github.com/zencart/zencart/assets/117885587/204120c1-a35c-44ad-aab1-0ab60ec5e5d6)

**Comment by drbyte** [comment]
> at first I liked the idea of dropping it but after few tests including Japanese version, I think keeping it is better. It will make it easier to use other languages and their date format.

Right, but did you ever customize that BEFORE this discussion? I don't think anybody actually customizes that, so isn't needed.

**Comment by piloujp** [comment]
> Right, but did you ever customize that BEFORE this discussion? I don't think anybody actually customizes that, so isn't needed.

Yes, I did for a while now in Japanese language pack. Date() has no international names (days, month, era...) support and can only be used with numbers in other language than English if you want a real local date.
In Japanese, I use a 'better date function' from PHP's intl library and use actual date() function for English or fall back in Japanese.

`		  echo ($_SESSION['language'] == 'japanese' AND function_exists('datefmt_create')) ? zen_set_local_calendar()->format(time()) : date(ADMIN_NAV_DATE_TIME_FORMAT);`

With lan.japanese.php containing this:

`    'ADMIN_NAV_DATE_TIME_FORMAT' => '%Y年 %m月 %d日 %A %X',`

Any way, this constant is only used in one place, which means that format string can be easily put in header.php code.
I was thinking that the advantage of keeping it, is that you can customize without touching header.php, but it is not true if you want to display days of week or months' in a language other than English.
Finally, I would say this constant can be dropped without any regret.

And for internationalization, it could only be achieved through IntlDateFormatter::format. But this is another problem.

**Comment by drbyte** [comment]
> for internationalization, it could only be achieved through IntlDateFormatter::format. But this is another problem.

`$zcDate->output()` uses `IntlDateFormatter` when `datefmt_create()` exists.

**Comment by drbyte** [comment]
I don't mind keeping the constant if it's going to be used.

And I don't mind dropping the use of `iconv()` if it's the source of problems.

I was hoping to drop the `if (is Windows) ? something : something for everyone else;` logic. But ... well ... Windows.

**Comment by piloujp** [comment]
> $zcDate->output() uses IntlDateFormatter when datefmt_create() exists.

Yes, but it is based on `strftime()` and don't use all possibilities of `IntlDateFormatter`, not to mention the bug with PHP 8.1.25 I think where everything crashed because `IntlDateFormatter` object was not instantiated properly.
I experimented in Japanese admin date with Japanese calendar based on Emperor Era. It can look fancy, but it is used everywhere in Japan, more on paper form than electronics ones due to computer limitations. I can't tell you how many times I had to fill my birthdate with this system...
A Zen Cart date system based on an `IntlDateFormatter` object would be nice! You would just have to make a new instance for each language, and no more limitation on how you display dates!

`I was hoping to drop the if (is Windows) ? something : something for everyone else; logic. But ... well ... Windows.`

This is your call.
There are probably still few ZC on Windows Server 2016 somewhere in the world but how many? And anyway, trouble in displaying admin won't be their major problem if they don't support UTF-8.
Generally, small business which are main ZC users don't use Windows Server, it is too heavy to maintain and too expensive.
Most of hosting companies are on Unix/Linux.
Considering all this, I would drop it...
